File: src_all/sources/MobileIMSDK4jXOpen/src/net/openmob/mobileimsdk/java/core/LocalUDPDataSender.java
/*
 * Copyright (C) 2017  即时通讯网(52im.net) & Jack Jiang.
 * The MobileIMSDK_X (MobileIMSDK v3.x) Project. 
 * All rights reserved.
 * 
 * > Github地址: https://github.com/JackJiang2011/MobileIMSDK
 * > 文档地址: http://www.52im.net/forum-89-1.html
 * > 即时通讯技术社区：http://www.52im.net/
 * > 即时通讯技术交流群：320837163 (http://www.52im.net/topic-qqgroup.html)
 *  
 * "即时通讯网(52im.net) - 即时通讯开发者社区!" 推荐开源工程。
 * 
 * LocalUDPDataSender.java at 2017-5-1 22:14:56, code by Jack Jiang.
 * You can contact author with jack.jiang@52im.net or jb2011@163.com.
 */
package net.openmob.mobileimsdk.java.core;

import java.net.DatagramSocket;
import java.net.InetAddress;

import net.openmob.mobileimsdk.java.ClientCoreSDK;
import net.openmob.mobileimsdk.java.conf.ConfigEntity;
import net.openmob.mobileimsdk.java.utils.Log;
import net.openmob.mobileimsdk.java.utils.UDPUtils;

import org.jdesktop.swingworker.SwingWorker;

import net.openmob.mobileimsdk.server.protocal.ProtocalFactory;
import net.openmob.mobileimsdk.server.protocal.ErrorCode;
import net.openmob.mobileimsdk.server.protocal.Protocal;

public class LocalUDPDataSender
{
	private final static String TAG = LocalUDPDataSender.class.getSimpleName();
	
	private static LocalUDPDataSender instance = null;
	
	public static LocalUDPDataSender getInstance()
	{
		if(instance == null)
			instance = new LocalUDPDataSender();
		return instance;
	}
	
	private LocalUDPDataSender()
	{
	}
	
	int sendLogin(String loginUserId, String loginToken, String extra)
	{
		byte[] b = ProtocalFactory.createPLoginInfo(loginUserId, loginToken, extra).toBytes();
		int code = send(b, b.length);
		if(code == 0)
		{
			ClientCoreSDK.getInstance().setCurrentLoginUserId(loginUserId);
			ClientCoreSDK.getInstance().setCurrentLoginToken(loginToken);
			ClientCoreSDK.getInstance().setCurrentLoginExtra(extra);
		}
		
		return code;
	}
	
	public int sendLoginout()
	{
		int code = ErrorCode.COMMON_CODE_OK;
		if(ClientCoreSDK.getInstance().isLoginHasInit())
		{
			byte[] b = ProtocalFactory.createPLoginoutInfo(
					ClientCoreSDK.getInstance().getCurrentLoginUserId()).toBytes();
			code = send(b, b.length);
			if(code == 0)
			{
				// do nothing
			}
		}
		ClientCoreSDK.getInstance().release();
		return code;
	}
	
	int sendKeepAlive()
	{
		byte[] b = ProtocalFactory.createPKeepAlive(
				ClientCoreSDK.getInstance().getCurrentLoginUserId()).toBytes();
		return send(b, b.length);
	}
		
	public int sendCommonData(String dataContentWidthStr, String to_user_id)
	{
		return sendCommonData(dataContentWidthStr, to_user_id, -1);
	}
	public int sendCommonData(String dataContentWidthStr, String to_user_id, int typeu)
	{
		return sendCommonData(dataContentWidthStr, to_user_id, null, typeu);
	}
	public int sendCommonData(String dataContentWidthStr, String to_user_id
			, String fingerPrint, int typeu)
	{
		return sendCommonData(dataContentWidthStr, to_user_id, true, fingerPrint, typeu);
	}
	public int sendCommonData(String dataContentWidthStr, String to_user_id
			, boolean QoS, String fingerPrint, int typeu)
	{
		return sendCommonData(ProtocalFactory.createCommonData(dataContentWidthStr
				, ClientCoreSDK.getInstance().getCurrentLoginUserId(), to_user_id, QoS, fingerPrint, typeu));
	}
	
	public int sendCommonData(Protocal p)
	{
		if(p != null)
		{
			byte[] b = p.toBytes();
			int code = send(b, b.length);
			if(code == 0)
			{
				if(p.isQoS() && !QoS4SendDaemon.getInstance().exist(p.getFp()))
					QoS4SendDaemon.getInstance().put(p);
			}
			return code;
		}
		else
			return ErrorCode.COMMON_INVALID_PROTOCAL;
	}
	private int send(byte[] fullProtocalBytes, int dataLen)
	{
		if(!ClientCoreSDK.getInstance().isInitialed())
			return ErrorCode.ForC.CLIENT_SDK_NO_INITIALED;
		
		DatagramSocket ds = LocalUDPSocketProvider.getInstance().getLocalUDPSocket();
		if(ds != null && !ds.isConnected())
		{
			try
			{
				if(ConfigEntity.serverIP == null)
				{
					Log.w(TAG, "【IMCORE】send数据没有继续，原因是ConfigEntity.server_ip==null!");
					return ErrorCode.ForC.TO_SERVER_NET_INFO_NOT_SETUP;
				}
				
				ds.connect(InetAddress.getByName(ConfigEntity.serverIP), ConfigEntity.serverUDPPort);
			}
			catch (Exception e)
			{
				Log.w(TAG, "【IMCORE】send时出错，原因是："+e.getMessage(), e);
				return ErrorCode.ForC.BAD_CONNECT_TO_SERVER;
			}
		}
		return UDPUtils.send(ds, fullProtocalBytes, dataLen) ? ErrorCode.COMMON_CODE_OK : ErrorCode.COMMON_DATA_SEND_FAILD;
	}
	
	public static abstract class SendCommonDataAsync extends SwingWorker<Integer, Object>
	{
		protected Protocal p = null;
		
		public SendCommonDataAsync(String dataContentWidthStr, String to_user_id)
		{
			this(dataContentWidthStr, to_user_id, null, -1);
		}
		public SendCommonDataAsync(String dataContentWidthStr, String to_user_id
				, int typeu)
		{
			this(dataContentWidthStr, to_user_id, null, typeu);
		}
		public SendCommonDataAsync(String dataContentWidthStr, String to_user_id
				, String fingerPrint, int typeu)
		{
			this(ProtocalFactory.createCommonData(dataContentWidthStr
					, ClientCoreSDK.getInstance().getCurrentLoginUserId(), to_user_id
					, true, fingerPrint, typeu));
		}

		public SendCommonDataAsync(Protocal p)
		{
			if(p == null)
			{
				Log.w(TAG, "【IMCORE】无效的参数p==null!");
				return;
			}
			this.p = p;
		}

		@Override
		protected Integer doInBackground()
		{
			if(p != null)
				return LocalUDPDataSender.getInstance().sendCommonData(p);//dataContentWidthStr, to_user_id);
			return -1;
		}
		
		@Override
		protected void done()
		{
			int code = -1;
			try
			{
				code = get();
			}
			catch (Exception e)
			{
				Log.w(TAG, e.getMessage());
			}
			
			onPostExecute(code);
		}

		protected abstract void onPostExecute(Integer code);
	}
	
	public static class SendLoginDataAsync extends SwingWorker<Integer, Object>
	{
		protected String loginUserId = null;
		protected String loginToken = null;
		protected String extra = null;

		public SendLoginDataAsync(String loginUserId, String loginToken)
		{
			this(loginUserId, loginToken, null);
		}

		public SendLoginDataAsync(String loginUserId, String loginToken, String extra)
		{
			this.loginUserId = loginUserId;
			this.loginToken = loginToken;
			this.extra = extra;
			
			ClientCoreSDK.getInstance().init();
		}

		@Override
		protected Integer doInBackground()
		{
			int code = LocalUDPDataSender.getInstance().sendLogin(
					this.loginUserId, this.loginToken, this.extra);
			return code;
		}

		@Override
		protected void done()
		{
			int code = -1;
			try
			{
				code = get();
			}
			catch (Exception e)
			{
				Log.w(TAG, e.getMessage());
			}
			
			onPostExecute(code);
		}
		
		protected void onPostExecute(Integer code)
		{
			if(code == 0)
			{
				LocalUDPDataReciever.getInstance().startup();
			}
			else
			{
				Log.d(TAG, "【IMCORE】数据发送失败, 错误码是："+code+"！");
			}
			
			fireAfterSendLogin(code);
		}
		
		protected void fireAfterSendLogin(int code)
		{
			// default do nothing
		}
	}
}

File: src_all/libs_src/MobileIMSDK4i_x_open/MobileIMSDK4i/Core/Impl/LocalUDPDataSender.m
//  ----------------------------------------------------------------------
//  Copyright (C) 2017  即时通讯网(52im.net) & Jack Jiang.
//  The MobileIMSDK_X (MobileIMSDK v3.x) Project.
//  All rights reserved.
//
//  > Github地址: https://github.com/JackJiang2011/MobileIMSDK
//  > 文档地址: http://www.52im.net/forum-89-1.html
//  > 即时通讯技术社区：http://www.52im.net/
//  > 即时通讯技术交流群：320837163 (http://www.52im.net/topic-qqgroup.html)
//
//  "即时通讯网(52im.net) - 即时通讯开发者社区!" 推荐开源工程。
//
//  如需联系作者，请发邮件至 jack.jiang@52im.net 或 jb2011@163.com.
//  ----------------------------------------------------------------------
//
//  LocalUDPDataSender.m
//  MibileIMSDK4i_X (MobileIMSDK v3.0 at Summer 2017)
//
//  Created by JackJiang on 14/10/27.
//  Copyright (c) 2017年 52im.net. All rights reserved.
//

#import "LocalUDPDataSender.h"
#import "ProtocalFactory.h"
#import "ClientCoreSDK.h"
#import "KeepAliveDaemon.h"
#import "CharsetHelper.h"
#import "QoS4SendDaemon.h"
#import "ErrorCode.h"
#import "GCDAsyncUdpSocket.h"
#import "LocalUDPSocketProvider.h"
#import "ConfigEntity.h"
#import "UDPUtils.h"
#import "CompletionDefine.h"

@implementation LocalUDPDataSender

static LocalUDPDataSender *instance = nil;


- (int) sendImpl_:(NSData *)fullProtocalBytes
{
    if(![[ClientCoreSDK sharedInstance] isInitialed])
        return ForC_CLIENT_SDK_NO_INITIALED;
    
    if(![ClientCoreSDK sharedInstance].localDeviceNetworkOk)
    {
        NSLog(@"【IMCORE】本地网络不能工作，send数据没有继续!");
        return ForC_LOCAL_NETWORK_NOT_WORKING;
    }

    GCDAsyncUdpSocket *ds = [[LocalUDPSocketProvider sharedInstance] getLocalUDPSocket];
    if(ds != nil && ![ds isConnected])
    {
        ConnectionCompletion observerBlock = ^(BOOL connectResult) {
            if(connectResult)
                [UDPUtils send:ds withData:fullProtocalBytes];
            else
            {
            }
        };
        [[LocalUDPSocketProvider sharedInstance] setConnectionObserver:observerBlock];

        NSError *connectError = nil;
        int connectCode = [[LocalUDPSocketProvider sharedInstance] tryConnectToHost:&connectError withSocket:ds completion:observerBlock];
        if(connectCode != COMMON_CODE_OK)
            return connectCode;
        else
            return COMMON_CODE_OK;
    }
    else
    {
        return [UDPUtils send:ds withData:fullProtocalBytes] ? COMMON_CODE_OK : COMMON_DATA_SEND_FAILD;
    }
}

- (void) putToQoS:(Protocal *)p
{
    if(p.QoS && ![[QoS4ReciveDaemon sharedInstance] exist:p.fp])
        [[QoS4ReciveDaemon sharedInstance] put:p];
}

+ (LocalUDPDataSender *)sharedInstance
{
    if (instance == nil)
    {
        instance = [[super allocWithZone:NULL] init];
    }
    return instance;
}

- (int) sendLogin:(NSString *)loginUserId withToken:(NSString *)loginToken
{
    return [self sendLogin:loginUserId withToken:loginToken andExtra:nil];
}

- (int) sendLogin:(NSString *)loginUserId withToken:(NSString *)loginToken andExtra:(NSString *)extra
{
    [[ClientCoreSDK sharedInstance] initCore];
    
    NSData *b = [[ProtocalFactory createPLoginInfo:loginUserId withToken:loginToken andExtra:extra] toBytes];
    int code = [self sendImpl_:b];
    if(code == 0)
    {
        [[ClientCoreSDK sharedInstance] setCurrentLoginUserId:loginUserId];
        [[ClientCoreSDK sharedInstance] setCurrentLoginToken:loginToken];
        [[ClientCoreSDK sharedInstance] setCurrentLoginExtra:extra];
    }
    
    return code;
}

- (int) sendLoginout
{
    int code = COMMON_CODE_OK;
    if([ClientCoreSDK sharedInstance].loginHasInit)
    {
        NSString *loginUserId = [ClientCoreSDK sharedInstance].currentLoginUserId;
        NSData *b = [[ProtocalFactory createPLoginoutInfo:loginUserId] toBytes];
        code = [self sendImpl_:b];
        if(code == 0)
        {
            [[KeepAliveDaemon sharedInstance] stop];
            [[ClientCoreSDK sharedInstance] setLoginHasInit:NO];
        }
    }
    
    [[ClientCoreSDK sharedInstance] releaseCore];
    
    return code;
}

- (int) sendKeepAlive
{
    NSString *currentLoginUserId = [[ClientCoreSDK sharedInstance] currentLoginUserId];
    NSData *b = [[ProtocalFactory createPKeepAlive: currentLoginUserId] toBytes];
    return [self sendImpl_:b];
}

- (int) sendCommonDataWithStr:(NSString *)dataContentWidthStr toUserId:(NSString *)to_user_id
{
    return [self sendCommonDataWithStr:dataContentWidthStr toUserId:to_user_id withTypeu:-1];
}

- (int) sendCommonDataWithStr:(NSString *)dataContentWidthStr toUserId:(NSString *)to_user_id withTypeu:(int)typeu
{
    NSString *currentLoginUserId = [[ClientCoreSDK sharedInstance] currentLoginUserId];
    Protocal *p = [ProtocalFactory createCommonData:dataContentWidthStr fromUserId:currentLoginUserId toUserId:to_user_id withTypeu:typeu];
    return [self sendCommonData:p];
}

- (int) sendCommonDataWithStr:(NSString *)dataContentWidthStr toUserId:(NSString *)to_user_id qos:(BOOL)QoS fp:(NSString *)fingerPrint withTypeu:(int)typeu
{
    NSString *currentLoginUserId = [[ClientCoreSDK sharedInstance] currentLoginUserId];
    Protocal *p = [ProtocalFactory createCommonData:dataContentWidthStr fromUserId:currentLoginUserId toUserId:to_user_id qos:QoS fp:fingerPrint withTypeu:typeu];
    return [self sendCommonData:p];
}

- (int) sendCommonData:(Protocal *)p
{
    @synchronized(self)
    {
        if(p != nil)
        {
            NSData *b = [p toBytes];
            int code = [self sendImpl_:b];
            if(code == 0)
            {
                [self putToQoS:p];
            }
            return code;
        }
        else
            return COMMON_INVALID_PROTOCAL;
    }
}

@end
